src/sqlfluff/core/templaters/jinja.py

```diff
@@ -326,8 +327,16 @@ def _slice_template(cls, in_str: str) -> Iterator[RawFileSlice]:
             "raw_end": "block",
             "raw_begin": "block",
         }
-        # https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.Environment.lex
-        for _, elem_type, raw in env.lex(in_str):
+        def lex(env, in_str):
+            # This is ugly. Jinja discards leading whitespace-only lines. We
+            # have to detect and add them back in order for SQLFluff's source
+            # position markers to align with the actual file.
+            m = re.search(r'^\s+\n', in_str)
+            if m:
+                yield None, "data", m.group(0)
+            # https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.Environment.lex
+            yield from env.lex(in_str)
+        for _, elem_type, raw in lex(env, in_str):
```

Comment:
Actually I found a real cheeky fix that seems to work for both #1321 and #854 now you've done the hard work and led the way:

```suggestion

        # Using Jinja whitespace stripping (e.g. `{#-` or `-%}`) breaks the
        # position markers between unlexed and lexed file. So let's ignore any
        # request to do that before lexing, by replacing '-' with '+'
        #
        # Note: '+' is the default, so shouldn't really be needed but we
        # explicitly state that to preserve the space for the missing '-' character
        # so it looks the same.
        in_str = in_str.replace('{%-', '{%+')
        in_str = in_str.replace('-%}', '+%}')
        in_str = in_str.replace('{#-', '{#+')
        in_str = in_str.replace('-#}', '+#}')

        for _, elem_type, raw in env.lex(env, in_str):
```

If so we should remove the `re` import above as no longer needed and also enhance this with other test cases (the example above I suggested and the one from #854).